kong/plugins/ddtrace/datadog_propagation.lua

```diff
@@ -0,0 +1,145 @@
+-- Propagation methods for Datadog APM: extraction and injection of datadog-specific request headers.
+
+local ffi = require "ffi"
+ffi.cdef[[
+unsigned long long int strtoull(const char *nptr, char **endptr, int base);
+]]
+
+local function id_to_string(id)
+  -- when concerted to a string, uint64_t values have ULL at the end of the string.
+  -- string.sub is used to remove the last 3 characters.
+  local str_id = tostring(id)
+  return string.sub(str_id, 1, #str_id - 3)
+end
+
+local function parse_uint64(str)
+    if not str then
+        return nil, "unable to parse, value is nil"
+    end
+    ffi.errno(0)
+    local parsed_str = ffi.C.strtoull(str, nil, 10)
+    local err = ffi.errno()
+    if err ~= 0 then
+        return nil, "unable to parse '" .. str .. "' as 64-bit number, errno=" .. err
+    end
+    -- TODO: check the entire string was consumed, instead of partially decoded
+    return parsed_str, nil
+end
+
+local function parse_dd_tags(s)
+    local tags = {}
+
+    for k,v in string.gmatch(s, "(_dd%.p%.[%w-._]+)=([%w-._]+)") do
+        if k ~= "_dd.p.upstream_services" then
+            tags[k] = v
+        end
+    end
+
+    return tags
+end
+
+local function encode_propagation_tags(tags)
+  local function startswith(s, pattern)
+    return s:sub(1, #pattern) == pattern
+  end
+
+  local result = nil
+  local first = true
+
+  for k,v in pairs(tags) do
+    if startswith(k, "_dd.p") then
+      local tag = k .. "=" .. v
+      if first then
+        result = tag
+        first = false
+      else
+        result = result .. "," .. tag
+      end
+    end
+  end
+
+  return result
+end
+
+local function extract_datadog(get_header, max_header_size)
+    local trace_id_value = get_header("x-datadog-trace-id")
+    if not trace_id_value then
+        -- no trace ID found, therefore create a new span
+        return nil, nil, nil, nil, nil, nil, nil
+    end
+    local trace_id, err = parse_uint64(trace_id_value)
+    if err then
+        -- tracing was desired but the value wasn't understood
+        return nil, nil, nil, nil, nil, nil, err
+    end
+
+    -- other headers are expected but aren't provided in all cases
+    local parent_id = parse_uint64(get_header("x-datadog-parent-id"))
+    local sampling_priority = tonumber(get_header("x-datadog-sampling-priority"))
+    local origin = get_header("x-datadog-origin")
+    local dd_tags = {}
+
+    if max_header_size > 0 then
+        local dd_tags_value = get_header("x-datadog-tags")
+        if dd_tags_value then
+            if #dd_tags_value > max_header_size then
+              dd_tags["_dd.propagation_error"] = "extract_max_size"
+              kong.log.warn("`x-datadog-tags` exceed the limit of " .. max_header_size .. " characters")
+            else
+              dd_tags, err = parse_dd_tags(dd_tags_value)
```

Comment:
From our discussion: Decide what kind of error handing / return style you want for `parse_dd_tags`. You probably will have to revisit the RFC that specified X-Datadog-Tags to know which sorts of parsing failures require the `_dd.propagation_error` annotation to be set.